### DISCUSSION TITLE
playing a sound more than once

### DISCUSSION THREAD

**Discussion by stuartsmith-nasstar** [pschatzmann/arduino-audio-tools#724] [Q&A] [answered]
In streams-memory_raw-i2s you should playing a raw file from memory, with most of the constuctors at the global level.

Once the copier.copy has finished playing/streaming the sound, how do I 'requeue' it for playing again ? i.e. the sort of use case where a button press plays a sound every time. I've had a look through the examples, wifi and discussions, but I'm drawing a blank. Sorry if this is something very simply. I need the copier object at global level so it can be repeatedly called in the loop, but  it seems that once it's did a 'copy' that is it.. so I need a new copier object ? Is there any chance someone could put a simple modified example here based on streams-memory_raw-i2s, but which plays the sound in a loop ? and or each time a function is called ?

**Comment by pschatzmann** [comment, marked as answer]
You should not look in the examples but in the class documentation.

Hint: There is a easy solution where you can request the object to loop or a more difficult one where you restart when it is finished...

**Reply by stuartsmith-nasstar** [reply]
sigh. where's the rolling eyes emoticon when u need it. But I suppose thanks for the cryptic reply. I eventually found out that you need to control the stream (for some reason). So here it's MemoryStream - set to loop, register a callback with the rewind method, then control it with flags. So for future searchers - hopefully that will prove useful to them without having to solve your riddles.

**Reply by pschatzmann** [reply]
Oh, I have expected that it works like in DynamicMemoryStream where setLoop(true) is the only thing needed.
I will open an issue...

**Reply by pschatzmann** [reply]
Just calling setLoop(true) is good enough! You can use the callback to be notified about the fact that it was rewinding, but this is optional...

**Reply by Benik3** [reply]
Interesting, but I don't see any info about setLoop in documentation:
![obrazek](https://user-images.githubusercontent.com/4532647/230488896-95f22aa8-8bc3-418b-be27-66a252bb2aad.png)

**Reply by pschatzmann** [reply]
[MemoryStream](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_memory_stream.html) has this method...

**Reply by Benik3** [reply]
Thanks. Anyway the setloop is "forced". Can you please point to more info about the "restart when finished"?
I'm working on one prop and I would like to make sound of "shoot". But because it's like machine gun the shots are repeated each cca 100ms and I need to trigger each individually (because of sync with HW).

**Reply by Benik3** [reply]
Got it! I just had to call begin() on the MemoryStream which resets it :)

**Reply by veitalmaphil** [reply]
hi @Benik3 do you need to play the first sound completely or can you play the sound shortened?
i'm trying to do the same thing, i have a shot-sound which is around 600ms or so and i want to force repeat it after 150ms. 
but it seems the sound needs to be played once completely and then it can be repeated as i want it. i also use begin().

maybe you can paste me a code snippet? or did you have this problem as well and know a solution?

**Reply by Benik3** [reply]
Hi, fortunately I didn't have to play the sound only partly (but I also though if it's possible to overcome some rare cases where it could be helpful). 
So I simply do:
``` c
 if (play_sound)
    {
        if (!copier.copy())
        {
            music.begin(); // reset stream
            play_sound = false;
        }
    }
```

I think that if you need to play the sound only partly, you have to "flush" the already filled buffer (if it's even possible).
The music.begin() just reset the stream from which the data are taken...

**Reply by veitalmaphil** [reply]
yes that is how i implemented this, and it works perfectly if the sound is played once fully. it was 6 months ago that i tried this, i just stumbled over this topic ... so thanks, i might try this soon again.
